File: week1/pset1/credit.c
#include<stdio.h>
#include<cs50.h>
#include<math.h>
int get_digits(long num, int n);
int find_length(long num);
int get_first_two_digits(long num,  int num_len);
int checksum(long cnum, int num_len);

int main(void){
    long cnum;
    int number;
    int clen;
    int two_digits;
    int total;

    cnum = get_long("Number: ");
    // printf("%ld\n\n", cnum);

    // number = get_digits(cnum, i);
    // printf("Number: %d\n", number);

    clen = find_length(cnum);
    printf("Length cnum: %d\n", clen);

    total = checksum(cnum, clen);
    printf("Total: %d\n", total);

    // Get first two digits of credit card number
    two_digits = get_first_two_digits(cnum, clen);
    printf("Two digits: %d\n", two_digits);

    // Veryfying type and validity
    if (total % 10 == 0){
        // printf("This is valid, brand...\n");

        // Check if VISA
        if (clen == 13 || clen == 16){
            if (two_digits / 10 == 4){
                printf("VISA\n");
            }

            // Check if MASTERCARD
            else if (two_digits >= 51 & two_digits <= 55) {
                printf("MASTERCARD\n");
            }
            else{
                printf("INVALID\n");
            }
        }

        // Check if AMERICAN EXPRESS AMEX
        else if (clen == 15){
            if (two_digits == 37 || two_digits == 34){
                printf("AMEX\n");
            }
            else{
                printf("INVALID\n");
            }
        }
        else{
            printf("INVALID\n");
        }
    }
    else{
        printf("INVALID\n");
    }
}



int get_digits(long num, int n){
    long temp;
    temp = num / (long) pow(10,n-1);
    temp = temp % 10;
    // % (int) (pow(10, (n+1))) / (int) pow(10, n);

    return temp;

}

int checksum(long cnum, int num_len){
    // Every other digits starts from second-to-last digit
    // int number;
    int sum1 = 0;
    int sum2 = 0;
    int total;

    for (int i=2; i<=num_len; i+=2){
        int num = get_digits(cnum, i);
        num = num*2;

        // Check if num is two digits, separate digits and add them
        int nlen = find_length(num);

        // If current number has two digits, add them digit-wise
        if (nlen == 2){
            int digit1 = num / 10;
            int digit2 = num % 10;
            sum1 += (digit1 + digit2);
        }
        else{
            sum1 += num;
        }
    }

    // printf("\n");

    // printf("Not every other digits\n");
    for (int i=1; i<=num_len; i+=2){
        int num = get_digits(cnum, i);
        sum2 += num;
        // printf("%d ", num);
        // printf("sum: %d", sum2);
        // printf("\n");
    }

    // printf("Sum1: %d\n", sum1);
    // printf("Sum2: %d\n", sum2);


    return sum1 + sum2;
}

int find_length(long num){
    int i = 0;
    while (num > 0){
        num = num / 10;
        i += 1;
    }
    return i;
}

int get_first_two_digits(long num,  int num_len){
    num = num / (long) pow(10,num_len-2);

    return num;
}

File: week6/pset6/credit/credit.py
from cs50 import get_string

def main():
    # Assume that input only contain numeric characters
    credit_num = list(input("Number: "))

    # Get credit card number
    credit_len = len(credit_num)

    # Check validity using luhn's algorithm
    valid = luhn_algo(credit_num)

    # Grab first two digits of credit card number
    first_two = int("".join(credit_num[:2]))

    # Find the credit card brand
    brand = credit_brand(valid, credit_len, first_two)
    print(brand)


def luhn_algo(credit_number):
    sum_digits = 0
    sum_digits = 0

    # Starting from second to last digit, decrement with interval of 2
    for i in range(len(credit_number)-2, -1, -2):

        # Multiply those number by 2
        num = int(credit_number[i]) * 2

        # Add two digits add them digits-wise
        if num // 10 > 0:
            sum_digits += num // 10
            sum_digits += num % 10
        else:
            sum_digits += num

    # Add those with number not multiplied by two, starting from the end (last digit)
    for i in range(len(credit_number)-1, -1, -2):
        # print(credit_number[i])
        sum_digits += int(credit_number[i])

    if sum_digits % 10 == 0:
        return True
    else:
        return False


def credit_brand(valid, credit_length, first_two_digits):
    # If not valid return INVALID
    if not valid:
        return "INVALID"

    # If credit card length is 16 it can be either VISA or MASTERCARD
    if (credit_length == 16):

        # VISA starts with 4
        if first_two_digits // 10 == 4:
            return "VISA"

        # First two digits of MASTER CARD can be 51, 52, 53, 54, 55
        elif (first_two_digits >= 51 and first_two_digits <= 55):
            return "MASTERCARD"

        else:
            return "INVALID"

    # VISA can also has length of 13 and starts with 4
    elif (credit_length == 13 and first_two_digits // 10 == 4):
        return "VISA"

    # AMERICAN EXPRESS has length of 15 and starts with 34 or 37
    elif (credit_length == 15 and (first_two_digits == 34 or first_two_digits == 37)):
        return "AMEX"

    else:
        return "INVALID"


if __name__ == "__main__":
    main()